src/Shared/NativeMethodsShared.cs

```diff
@@ -470,6 +503,47 @@ public SystemInformationData()
             }
         }
 
+        public unsafe static int GetPhysicalCoreCount()
+        {
+            uint len = 0;
+            const int ERROR_INSUFFICIENT_BUFFER = 122;
+
+            if (!GetLogicalProcessorInformationEx(LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup, IntPtr.Zero, ref len) &&
+                Marshal.GetLastWin32Error() == ERROR_INSUFFICIENT_BUFFER)
+            {
+                // Allocate that much space
+                var buffer = new byte[len];
+                fixed (byte* bufferPtr = buffer)
+                {
+                    // Call GetLogicalProcessorInformationEx with the allocated buffer
+                    if (GetLogicalProcessorInformationEx(LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup, (IntPtr)bufferPtr, ref len))
+                    {
+                        // Walk each SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX in the buffer, where the Size of each dictates how
+                        // much space it's consuming.  For each group relation, count the number of active processors in each of its group infos.
+                        int processorCount = 0;
+                        byte* ptr = bufferPtr, endPtr = bufferPtr + len;
+                        while (ptr < endPtr)
+                        {
+                            var current = (SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*)ptr;
+                            if (current->Relationship == LOGICAL_PROCESSOR_RELATIONSHIP.RelationGroup)
+                            {
+                                PROCESSOR_GROUP_INFO* groupInfo = &current->Group.GroupInfo;
+                                int groupCount = current->Group.ActiveGroupCount;
+                                for (int i = 0; i < groupCount; i++)
+                                {
+                                    processorCount += (groupInfo + i)->ActiveProcessorCount;
```

Comment:
Add a check that this dereference is not overrunning the containing `buffer`, just in case.

Reply:
I'm not quite sure about the kind of check you have in mind here?
```
byte* ptr = bufferPtr, endPtr = bufferPtr + len;
while (ptr < endPtr)
```
This seems correct from what I can tell.
What check would you add to `processorCount += (groupInfo + i)->ActiveProcessorCount;`?

Reply:
If `groupCount` is a bogus number, `(groupInfo + i)` could be pointing outside of the `bufferPtr` <-> `endPtr` block. I believe it's a good security practice to not trust the data and check that all memory accesses are within the bounds when using unsafe C# code.

Strictly speaking, this should also include

Reply:
Since this code was directly taken from dotnet/runtime unmodified, and has been pushed to production, I'm inclined to believe it works and does not need additional checks. I'll add the checks since you prefer though.

Reply:
Added some checks https://github.com/dotnet/msbuild/pull/5625/commits/4223cc2e0c875c527afaf5f67a255c2995c4ebb1, please feel free to let me know if that's not what you had in mind.

Reply:
These checks will now just ignore invalid data which is, to our understanding, logically impossible. Should they not be asserting or throwing?